conjure-go-client/httpclient/body.go

```diff
@@ -0,0 +1,133 @@
+// Use this function when the body should be read only once (e.g. it is a response body stream from another request).
+//
+// The body's Close() method will be called when the request is completed. To disable, wrap the value in io.NopCloser.
+//
+// The input argument must be a function of one of the following types:
+//   - func() io.ReadCloser                 // Returns the body and no error
+//   - func() (io.ReadCloser, error)        // Returns the body and an error
+//   - func() (io.ReadCloser, int64, error) // Returns the body, content length, and an error
+func RequestBodyStreamOnce[T requestBodyStreamInput](input T) RequestBody {
+	return requestBodyFunc(func() (contentLen int64, body io.ReadCloser, getBody func() (io.ReadCloser, error), err error) {
+		switch v := any(input).(type) {
+		default:
+			// Cases below MUST be exhaustive of the generic type!
+			return 0, nil, nil, fmt.Errorf("httpclient.RequestBodyStreamOnce: unexpected type %T", v)
+		case nil:
+			return 0, nil, nil, nil
+		case func() io.ReadCloser:
+			return -1, v(), nil, nil
+		case func() (io.ReadCloser, error):
+			body, err = v()
+			return -1, body, v, err
+		case func() (io.ReadCloser, int64, error):
+			getBody = func() (io.ReadCloser, error) {
+				b, _, e := v()
+				return b, e
+			}
+			body, contentLen, err = v()
+			return contentLen, body, getBody, err
+		}
+	})
+}
+
+// RequestBodyStreamWithReplay sets the *http.Request Body and GetBody fields for upload.
+//
+// The GetBody field is set to a function that returns the same io.ReadCloser. The http.Transport will be able to replay the request body
+// when a request is redirected. Use this function when the body reader can be recreated multiple times.
+//
+// The body's Close() method will be called when the request is completed. To disable, wrap the value in io.NopCloser.
+//
+// The input argument must be a function of one of the following types:
+//   - func() io.ReadCloser                 // Returns the body and no error
+//   - func() (io.ReadCloser, error)        // Returns the body and an error
+//   - func() (io.ReadCloser, int64, error) // Returns the body, content length, and an error
+func RequestBodyStreamWithReplay[T requestBodyStreamInput](input T) RequestBody {
+	return requestBodyFunc(func() (contentLen int64, body io.ReadCloser, getBody func() (io.ReadCloser, error), err error) {
+		switch v := any(input).(type) {
+		default:
+			// Cases below MUST be exhaustive of the generic type!
+			return 0, nil, nil, fmt.Errorf("httpclient.RequestBodyStreamWithReplay: unexpected type %T", v)
+		case nil:
+			return 0, nil, nil, nil
+		case func() io.ReadCloser:
+			body = v()
+			getBody = func() (io.ReadCloser, error) {
+				return v(), nil
+			}
+			return -1, body, getBody, nil
+		case func() (io.ReadCloser, error):
+			body, err = v()
+			return -1, body, v, err
+		case func() (io.ReadCloser, int64, error):
```

Comment:
💯  the lack of correct content length headers on conjure requests where the body is type `binary` has been quite painful, seem like this would give consumers a chance to fix that.